### ISSUE TITLE
bad order book sorting

### ISSUE THREAD

**Issue by Pmant** [ccxt/ccxt#1364]
Order books are not sorted by price on some exchanges, see example. May be related to #1263.

- OS: Windows 10
- Programming Language version: Node.js 8.9.4
- CCXT version: 1.10.775
- Exchange: Cryptopia, Gate.io (and more?)
- Method: fetchL2OrderBook

```javascript
const ccxt = require('ccxt');

const cryptopia  = new ccxt.cryptopia({
    timeout: 30000
});

(async () => {
    await cryptopia.loadMarkets();
    let orderBook = await cryptopia.fetchL2OrderBook('BTC/USDT', {});
    console.log(orderBook);
})();

{ bids:
   [ [ 10150, 0.97979069 ],
     [ 10155, 0.11307884 ],
     [ 10200, 0.03727139 ],
     [ 10210, 0.001 ],
     [ 10250, 0.01759893 ],
     [ 10251, 0.22204858 ],
     [ 10300, 0.0239395 ],
     [ 10310, 0.001 ],
     [ 10350, 0.01222226 ],
     [ 10351, 0.1 ],
     [ 10400, 0.05135667 ],
     [ 10460, 0.06331433 ],
     [ 10500, 0.34749287 ],
     [ 10502, 0.42470445 ],
     [ 10503, 0.3 ],
     [ 10504, 0.90526149 ],
     [ 10505, 0.36 ],
     [ 10506, 0.36 ],
     [ 10507, 0.36 ],
     [ 10508, 0.36 ],
     [ 10509, 0.1743538 ],
     [ 10557, 0.0025 ],
     [ 10580, 0.09406456 ],
     [ 10588, 0.02757646 ],
     [ 10598, 0.00565015 ],
     [ 10600, 0.04962176 ],
     [ 10640, 0.0050124 ],
     [ 10650, 0.01382 ],
     [ 10700, 0.38987445 ],
     [ 10780, 0.23843951 ],
     [ 10800, 0.01025398 ],
     [ 10820, 0.01483021 ],
     [ 10830, 0.40896441 ],
     [ 10880, 0.02091405 ],
     [ 11000, 1.9959412 ],
     [ 11001, 1 ],
     [ 11002, 0.02246568 ],
     [ 11010, 0.01589902 ],
     [ 11118, 0.0512264 ],
     [ 11138, 0.28211238 ],
     [ 11250, 0.26187625 ],
     [ 11265, 0.5 ],
     [ 11300, 0.12480967 ],
     [ 11312, 0.0001 ],
     [ 11313, 0.0001 ],
     [ 11314, 0.0001 ],
     [ 11399, 0.005 ],
     [ 11420, 0.09972585 ],
     [ 11420.00000001, 0.0003496 ],
     [ 11313.000022, 0.02646523 ],
     [ 11312.00000001, 0.0263491 ],
     [ 11311.00400001, 0.03676902 ],
     [ 11311.004, 0.00473789 ],
     [ 11311.00030001, 0.07899997 ],
     [ 11309.19500003, 0.23592715 ],
     [ 11301.00000003, 0.02228129 ],
     [ 11297.55, 0.00378223 ],
     [ 11264.35, 0.01339363 ],
     [ 11254.858506, 0.00171848 ],
     [ 11116.96100012, 0.19791785 ],
     [ 11116.96100006, 0.00675837 ],
     [ 11116.96100005, 0.00720429 ],
     [ 11101.47778, 0.00018016 ],
     [ 11040.00010001, 0.02165371 ],
     [ 11040.00000005, 0.00540885 ],
     [ 11027.25000009, 0.00010883 ],
     [ 11024.35881551, 0.06336876 ],
     [ 11001.00000001, 0.05176689 ],
     [ 10876.68583419, 0.00025012 ],
     [ 10863.61077761, 0.00009206 ],
     [ 10863.00000003, 0.00900411 ],
     [ 10830.00000001, 0.0334418 ],
     [ 10811.02941185, 0.00012092 ],
     [ 10800.58600005, 0.03872494 ],
     [ 10800.00000004, 0.00184816 ],
     [ 10768.43344, 0.00018573 ],
     [ 10725.9196, 0.17305985 ],
     [ 10689.90000001, 0.04649025 ],
     [ 10650.59880157, 0.0000939 ],
     [ 10599.04844299, 0.00013435 ],
     [ 10445.38044, 0.00019148 ],
     [ 10441.76353095, 0.00009577 ],
     [ 10401.96545549, 0.03615684 ],
     [ 10391.22396371, 0.00014927 ],
     [ 10389.99999996, 0.06879354 ],
     [ 10380.00000008, 0.00622679 ],
     [ 10351.00050001, 0.06210099 ],
     [ 10351.0005, 0.64888868 ],
     [ 10348.52326704, 0.01928778 ],
     [ 10307.3547366, 0.11494273 ],
     [ 10300.15, 0.0487605 ],
     [ 10250.00000001, 0.13350186 ],
     [ 10244.92982464, 0.01948281 ],
     [ 10237.02306955, 0.00009769 ],
     [ 10200.99, 0.04320413 ],
     [ 10200.6035, 0.00611818 ],
     [ 10199.9999, 0.24254463 ],
     [ 10190.49999995, 0.00447722 ],
     [ 10187.47447422, 0.00016585 ],
     [ 10132.01903, 0.0001974 ] ],
  asks:
   [ [ 11492, 0.00348766 ],
     [ 11495, 0.04217789 ],
     [ 11498, 0.00556574 ],
     [ 11499, 0.00077778 ],
     [ 11500, 0.76672586 ],
     [ 11526, 0.00521605 ],
     [ 11527, 0.01 ],
     [ 11530, 0.0443 ],
     [ 11550, 0.42290939 ],
     [ 11551, 0.05 ],
     [ 11570, 0.002 ],
     [ 11596, 0.1 ],
     [ 11599, 0.07450138 ],
     [ 11600, 0.01493182 ],
     [ 11640, 0.00086871 ],
     [ 11650, 0.05656736 ],
     [ 11651, 0.05 ],
     [ 11670, 0.00062489 ],
     [ 11680, 0.03114414 ],
     [ 11683, 0.62501413 ],
     [ 11700, 2.4583732 ],
     [ 11745, 0.03425655 ],
     [ 11747, 0.00009805 ],
     [ 11749, 0.026547 ],
     [ 11750, 1.37559163 ],
     [ 11799, 0.12304789 ],
     [ 11800, 1.49735191 ],
     [ 11810, 0.004 ],
     [ 11820, 0.00847719 ],
     [ 11825, 0.00122152 ],
     [ 11838, 0.04137581 ],
     [ 11844, 0.01 ],
     [ 11848, 0.01405811 ],
     [ 11850, 0.06670701 ],
     [ 11851, 0.03179992 ],
     [ 11860, 0.04418132 ],
     [ 11900, 3.18013741 ],
     [ 11923, 0.01039148 ],
     [ 11950, 0.05304889 ],
     [ 11951, 0.05 ],
     [ 11955, 0.00408939 ],
     [ 11975, 0.01734637 ],
     [ 11997, 0.15340675 ],
     [ 11998, 0.05 ],
     [ 11999, 0.92987566 ],
     [ 12000, 7.64109408 ],
     [ 12050, 3 ],
     [ 12100, 3.01474924 ],
     [ 12150, 3.001 ],
     [ 12200, 3.04747614 ],
     [ 11491.99999999, 0.00086808 ],
     [ 11496.99799999, 0.00142819 ],
     [ 11505.219362, 0.00169618 ],
     [ 11526.99999, 0.00143834 ],
     [ 11530.86908993, 0.35517465 ],
     [ 11534.3220674, 0.00162015 ],
     [ 11549.6698935, 0.0000592 ],
     [ 11556.57550001, 0.0269454 ],
     [ 11577.777, 0.00396134 ],
     [ 11577.8702422, 0.00680152 ],
     [ 11630.00000001, 0.00019191 ],
     [ 11662.58461015, 0.000125 ],
     [ 11684.49128828, 0.00987221 ],
     [ 11692.17489747, 0.00010264 ],
     [ 11699.99999999, 0.005 ],
     [ 11700.00000004, 0.02559144 ],
     [ 11744.99999999, 0.00892224 ],
     [ 11759.99999997, 0.01057554 ],
     [ 11762.21024305, 0.00103329 ],
     [ 11781.78872738, 0.01 ],
     [ 11798.78603, 0.09597138 ],
     [ 11799.99999999, 0.00242281 ],
     [ 11800.99899998, 0.10313635 ],
     [ 11852.72350008, 0.00098297 ],
     [ 11854.9999997, 0.00757817 ],
     [ 11860.67930001, 0.18656542 ],
     [ 11863.0831643, 0.00009659 ],
     [ 11899.99999, 0.000506 ],
     [ 11899.99999019, 0.05 ],
     [ 11900.59654096, 0.000125 ],
     [ 11910.00000001, 0.49752366 ],
     [ 11930.7907117, 0.00011404 ],
     [ 11935.5, 0.002 ],
     [ 11948.52941176, 0.00009207 ],
     [ 11982.7, 0.012709 ],
     [ 11988.7, 0.00350453 ],
     [ 11998.7, 0.05546398 ],
     [ 11998.999997, 0.02 ],
     [ 11999.99999999, 0.00021496 ],
     [ 12044.81927711, 0.00047632 ],
     [ 12071.4, 0.01018335 ],
     [ 12101.11649418, 0.01154818 ],
     [ 12104.6971479, 0.00075 ],
     [ 12112.43394189, 0.00011384 ],
     [ 12143.46585812, 0.000125 ],
     [ 12152.73833671, 0.00009158 ],
     [ 12163.69694, 0.00016951 ],
     [ 12171.53688525, 0.00008754 ],
     [ 12174.27623642, 0.00012671 ],
     [ 12192.37695077, 0.0001023 ] ],
  timestamp: 1516359729911,
  datetime: '2018-01-19T11:02:09.911Z' }
```

**Comment by Deathamns** [comment]
Yes, it's related to that issue. Unfortunately the `aggregate` function uses methods that are not safe for sorting.
One option is to revert sorting for `fetchL2OrderBook`, but I think fixing the `aggregate` function would be better. I'll try to send a pull request.
I guess this is a JS specific issue.

**Comment by mkutny** [comment]
@Deathamns , +1 for fixing `aggregate`.

It was so painful that I had to make a quick-fix patch for myself, to fix the immediate issue and then investigate further.

Actually, `forEach` preserves order, it then gets lost in `results` object.

I'll close my PR when you'll submit yours, but I'd be grateful if you cherry-pick new testcases that cover this issue.

**Comment by Deathamns** [comment]
It's `Object.keys`, as in JavaScript the key order is not guaranteed.

```js
let x = {'1': 1, '1.1': 1.1, '2.2': 2.2, '2': 2};
Object.keys(x);
// Array [ "1", "2", "1.1", "2.2" ]
```

**Comment by Deathamns** [comment]
The fix is easy for the `aggregate` function, however it expects an already sorted array, which it will get from fetchOrderBook, and we don't even use it anywhere else, so I won't send a PR, I'll let @kroitor to decide if he wants this.

```js
const aggregate = function (bidasks) {
    let result = [];
    let last = [0, 0];
    for (let [price, volume] of bidasks) {
        volume = parseFloat (volume)
        if (volume > 0) {
            price = parseFloat (price)
            if (price === last[0]) {
                last[1] += volume
            } else {
                last = [price, volume]
                result.push (last)
            }
        }
    }
    return result
}
```

**Comment by kroitor** [comment]
Hi! 
This is fixed my @mkutny in #1370. Please update ccxt to version 1.10.779+.
Let us know if you have any other difficulties with it. Thx!
